jib-core/src/main/java/com/google/cloud/tools/jib/frontend/MainClassFinder.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright 2018 Google LLC. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.jib.frontend;
+
+import java.io.IOException;
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import javax.annotation.Nullable;
+
+/** Class used for inferring the main class in an application. */
+public class MainClassFinder {
+
+  /** Helper class for loading a .class file. */
+  private static class ClassFileLoader extends ClassLoader {
+
+    private Path classFile;
+
+    private ClassFileLoader(Path classFile) {
+      this.classFile = classFile;
+    }
+
+    @Nullable
+    @Override
+    public Class findClass(String name) {
+      try {
+        byte[] bytes = Files.readAllBytes(classFile);
+        return defineClass(name, bytes, 0, bytes.length);
+      } catch (IOException | ClassFormatError ignored) {
+        // Not a valid class file
+        return null;
+      }
+    }
+  }
+
+  /**
+   * Searches for a class containing a main method given an absolute root directory.
+   *
+   * @return the name of the class if one is found, null if no class is found.
+   * @throws IOException if searching/reading files fails.
+   * @throws MultipleClassesFoundException if more than one valid main class is found.
+   */
+  @Nullable
+  public static String findMainClass(String rootDirectory)
+      throws MultipleClassesFoundException, IOException {
+    // Make sure rootDirectory is valid
+    if (!Files.exists(Paths.get(rootDirectory)) || !Files.isDirectory(Paths.get(rootDirectory))) {
+      return null;
+    }
+
+    String className = null;
+    try (Stream<Path> pathStream = Files.walk(Paths.get(rootDirectory))) {
+      // Get all .class files
+      List<Path> classFiles =
+          pathStream
+              .filter(Files::isRegularFile)
+              .filter(path -> path.toString().endsWith(".class"))
+              .collect(Collectors.toList());
+
+      for (Path classFile : classFiles) {
+        // Convert filename (rootDir/path/to/ClassName.class) to class name (path.to.ClassName)
+        String name = classFile.toAbsolutePath().toString();
+        name = name.substring(rootDirectory.length() + 1);
+        name = name.replace('/', '.').replace('\\', '.');
```

Comment:
`\` is actually a valid character for a filename in Unix. A safer way to do this conversion would be to just get the parts of the `Path` by iterating over it.